plugins/carbon_black_cloud/icon_carbon_black_cloud/tasks/monitor_alerts/task.py

```diff
@@ -0,0 +1,301 @@
+        :param now: current time to determine the start time for each CB type.
+        :param saved_state: dictionary of state values held.
+        :return: two string formatted times to apply to our alert and observation queries.
+        """
+        # take a copy of state so this logic will need to happen again if an exception occurs
+        state = saved_state.copy()
+
+        log_msg = ""
+        for cb_type_time in [LAST_ALERT_TIME, LAST_OBSERVATION_TIME]:
+            saved_time = state.get(cb_type_time)
+            if not saved_time:
+                log_msg += f"No {cb_type_time} within state. "
+                custom_timings = custom_config.get(cb_type_time, {})
+                custom_date = custom_timings.get("date")
+                custom_minutes = custom_timings.get("minutes", DEFAULT_LOOKBACK)
+                start = datetime(**custom_date) if custom_date else (now - timedelta(minutes=custom_minutes))
+                state[cb_type_time] = start.strftime(TIME_FORMAT)
+            else:
+                # check if we have held the TS beyond our max lookback
+                lookback_days = custom_config.get(f"{cb_type_time}_days", MAX_LOOKBACK)
+                default_date_lookback = now - timedelta(days=lookback_days)  # if not passed from CPS create on the fly
+                custom_lookback = custom_config.get(f"max_{cb_type_time}", {})
+                comparison_date = datetime(**custom_lookback) if custom_lookback else default_date_lookback
+                comparison_date = comparison_date.replace(tzinfo=timezone.utc).strftime(TIME_FORMAT)
+                if comparison_date > saved_time:
+                    self.logger.info(f"Saved time ({saved_time}) exceeds cut off, moving to ({comparison_date}).")
+                    state[cb_type_time] = comparison_date
+
+        alerts_start = state.get(LAST_ALERT_TIME)
+        observation_start = state.get(LAST_OBSERVATION_TIME)
+
+        self.logger.info(
+            f"{log_msg}Applying the following start times: alerts='{alerts_start}' "
+            f"and observations='{observation_start}'"
+        )
+        return alerts_start, observation_start
+
+    def _check_if_job_time_exceeded(self, job_start_time: str) -> bool:
+        """
+        Jobs can only run within CB for a maximum of 3 minutes, allow a time frame of 4 minutes to complete otherwise we
+        then we can assume that the job has been cancelled due to high usage on their platform and there will be
+        a difference of 1 between the contacted and completed values and we should parse the available observations.
+        """
+        job_cut_off = (self._get_current_time() - timedelta(minutes=4)).strftime(TIME_FORMAT)
+
+        if job_cut_off > job_start_time:
+            self.logger.info(
+                f"Job has exceeded max run time. Started at {job_start_time}. Parsing available results..."
```

Comment:
(Suggestion): Is there a job id or similar we could log here for troubleshooting?

Reply:
We look the job ID just before this, but added into this log line as well - 
```
{"message": "Get observation results from saved ID: example-sqs", "request_path": "/api/v1/tasks/monitor_alerts", "level": "info", "timestamp": "2024-05-01 13:36.06"}
{"message": "Job (example-sqs) has exceeded max run time. Started at 2024-04-01T13:34:03.626Z. Parsing available results...", "request_path": "/api/v1/tasks/monitor_alerts", "level": "info", "timestamp": "2024-05-01 13:36.08"}
{"message": "Expecting to dedupe a max of 1 based on previously stored hashes.", "request_path": "/api/v1/tasks/monitor_alerts", "level": "info", "timestamp": "2024-05-01 13:36.08"}
```